### PR TITLE
cmd/containerboot: add support for setting funnel TCP portforward

### PR DESCRIPTION
**Pull request by maisem**
WIP

Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: cmd/containerboot/main.go [hunk 1]
```diff
@@ -90,8 +90,31 @@ func main() {
 		AuthOnce:        defaultBool("TS_AUTH_ONCE", false),
 		Root:            defaultEnv("TS_TEST_ONLY_ROOT", "/"),
 	}
+	funnelForwardPorts := strings.Split(defaultEnv("TS_FUNNEL_TCP_PORTFORWARD", ""), ",")
```
Comment (danderson) [inline cmd/containerboot/main.go:93]: I'd probably call this just "TS_FUNNEL_TCP", to reduce typing. Also add the envvar to the big comment at the top that lists all the available options.

File: cmd/containerboot/main.go [hunk 2]
```diff
@@ -305,6 +324,35 @@ authLoop:
...
+		TCP:         make(map[uint16]*ipn.TCPPortHandler),
+	}
+	for f, t := range cfg.FunnelTCPPorts {
+		sc.TCP[f] = &ipn.TCPPortHandler{
+			TCPForward:   fmt.Sprintf("%s:%d", cfg.ProxyTo, t),
+			TerminateTLS: cd,
+		}
+		sc.AllowFunnel[ipn.HostPort(fmt.Sprintf("%s:%d", cd, f))] = true
+	}
+	return client.SetServeConfig(ctx, sc)
```
Comment (danderson) [inline cmd/containerboot/main.go:353]: How do we know when funnel is working? Ideally the operator would like some signal that pod startup has completed correctly, so it doesn't advertise the ingress hostname until it's actually serving.

File: cmd/containerboot/main.go [hunk 3]
```diff
@@ -305,6 +324,35 @@ authLoop:
 	}
 }
 
+func configureForwarding(ctx context.Context, client *tailscale.LocalClient, cfg *settings) error {
+	if cfg.ProxyTo == "" {
```
Comment (danderson) [inline cmd/containerboot/main.go:328]: I don't think this is right for all scenarios, it's valid to configure funnel without ProxyTo if you're running the container as a sidecar. I think funnel needs to work in either case?

**Review by danderson — COMMENTED** [review]
I don't think I like this as-is, especially the various "if kubernetes" branches in ipnlocal. We already have a bunch of these branches throughout our code and the magically different behavior makes it hard to maintain stuff :(

Maybe it's time to graduate the store API from plain key/value to structured Get/Set for specific data, so that the underlying implementations cna handle all this weird differential storage we're doing.

**Comment by DentonGentry** [comment]
Recommend the commit message state that it Updates https://github.com/tailscale/tailscale/issues/6468
(or Fixes, as the case may be)

**Pull request closed without merging**